ad/internal/winrmhelper/winrm_gmsa.go

```diff
@@ -0,0 +1,464 @@
+	}
+
+	cmds = append(cmds, fmt.Sprintf("-TrustedForDelegation $%t", g.TrustedForDelegation))
+
+	result, err := RunWinRMCommand(client, cmds, true, false, execLocally)
+	if err != nil {
+		return "", err
+	}
+
+	if result.ExitCode != 0 {
+		log.Printf("[DEBUG] stderr: %s\nstdout: %s", result.StdErr, result.Stdout)
+		if strings.Contains(result.StdErr, "AlreadyExists") {
+			return "", fmt.Errorf("there is another gmsa named %q", g.Name)
+		}
+		return "", fmt.Errorf("command New-ADgmsa exited with a non-zero exit code %d, stderr: %s", result.ExitCode, result.StdErr)
+	}
+
+	gmsa, err := unmarshallGmsa([]byte(result.Stdout))
+
+	if err != nil {
+		return "", fmt.Errorf("error while unmarshalling gmsa json document: %s", err)
+	}
+
+	return gmsa.GUID, nil
+}
+
+// ModifyGmsa updates the AD gmsa's details based on what's changed in the resource.
+func (g *Gmsa) ModifyGmsa(d *schema.ResourceData, client *winrm.Client, execLocally bool) error {
+	log.Printf("Modifying gmsa: %q", g.Name)
+	strKeyMap := map[string]string{
+		"expiration":       "AccountExpirationDate",
+		"sam_account_name": "SamAccountName",
+		"display_name":     "DisplayName",
+		"description":      "Description",
+		"dns_host_name":    "DNSHostName",
+		"home_page":        "HomePage",
+		"name":             "Name",
+	}
+
+	cmds := []string{fmt.Sprintf("Set-ADServiceAccount -Identity %q", g.GUID)}
+
+	for k, param := range strKeyMap {
+		if d.HasChange(k) {
+			value := SanitiseTFInput(d, k)
+			if value == "" {
+				value = "$null"
+			} else {
+				value = fmt.Sprintf(`"%s"`, value)
+			}
+			cmds = append(cmds, fmt.Sprintf(`-%s %s`, param, value))
+		}
+	}
+
+	boolKeyMap := map[string]string{
+		"delegated":              "AccountNotDelegated ",
+		"enabled":                "Enabled",
+		"trusted_for_delegation": "TrustedForDelegation",
+	}
+
+	for k, param := range boolKeyMap {
+		if d.HasChange(k) {
+			value := d.Get(k).(bool)
+			cmds = append(cmds, fmt.Sprintf("-%s $%t", param, value))
+		}
+	}
+
+	if len(cmds) > 1 {
+		result, err := RunWinRMCommand(client, cmds, false, false, execLocally)
+		if err != nil {
+			return err
+		}
+		if result.ExitCode != 0 {
+			log.Printf("[DEBUG] stderr: %s\nstdout: %s", result.StdErr, result.Stdout)
+			return fmt.Errorf("command Set-ADServiceAccount exited with a non-zero exit code %d, stderr: %s", result.ExitCode, result.StdErr)
+		}
+	}
+
+	if d.HasChange("service_principal_names") {
+		sp := []string{}
+		sprinc := d.Get("service_principal_names").(*schema.Set)
+		for _, s := range sprinc.List() {
+			if s == "" {
+				continue
+			}
+			sp = append(sp, s.(string))
+		}
+		cmds := []string{fmt.Sprintf("Set-ADServiceAccount -Identity %q -ServicePrincipalNames $null", g.GUID)}
+		if len(sp) > 0 {
+			spnlist := strings.Join(sp, ",")
+			log.Printf("[DEBUG] SPN list: %s", spnlist)
+			cmds = append(cmds, fmt.Sprintf("; Set-ADServiceAccount -Identity %q -ServicePrincipalNames @{Add=%q}", g.GUID, spnlist))
```

Comment:
This is definitely a simpler way to handle additions _and_ deletetions without having to compare lists 
etc. My only concern here is that if this command fails for any reasons then the ServicePrincipalNames 
attribute will be left empty, which could potentially cause operational issues.

What do you think?

Reply:
Indeed it would be better to update only what is needed for all parameters that have a list (SPN, principals etc) but the thing is that I'm not super comfortable (yet) to do this part.
I would need to look at what you did in the membership resource.
I'm afraid it's gonna take me a bit of time to achieve it but I fully agree it's the smoothest approach to avoid problems.

Reply:
In this [example](https://github.com/hashicorp/terraform-provider-ad/blob/a8e3f9382cac190bf4dd13988771422838ba60b4/ad/internal/winrmhelper/winrm_group_membership.go#L117) I call [diffGroupMemberLists](https://github.com/hashicorp/terraform-provider-ad/blob/a8e3f9382cac190bf4dd13988771422838ba60b4/ad/internal/winrmhelper/winrm_group_membership.go#L33). 

In that function I am passing two lists as arguments, the first one is the list of members as expressed in the configuration and the second is a list of members as retrieved from AD. It then compares one to the other and returns two new lists, a list of members to add and a list of members to remove.

It's not exactly performant, but given that the lists are relatively small it will do what we need.